pkg/internal/infrastructure/helper.go

```diff
@@ -0,0 +1,122 @@
+
+	kubevirtapi "github.com/gardener/gardener-extension-provider-kubevirt/pkg/apis/kubevirt"
+	"github.com/gardener/gardener-extension-provider-kubevirt/pkg/apis/kubevirt/helper"
+
+	"github.com/gardener/gardener/pkg/utils"
+
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	TenantNetworkAnnotation = "tenant.network.infra.gardener.io"
+)
+
+// CreateNetworkAttachmentDefinitionIfNotExist returns the GetNeworkAttachmentDefinition object matching the given GetNeworkAttachmentDefinitionRef.
+// If no existing NetworkAttachmentDefinition found, create and return one using CNI configs.
+func CreateNetworkAttachmentDefinitionIfNotExist(ctx context.Context, c client.Client, secretRef corev1.SecretReference, nadRef []*kubevirtapi.NetworkAttachmentDefinitionRef) ([]networkv1.NetworkAttachmentDefinition, error) {
+	if len(nadRef) == 0 {
+		return nil, nil
+	}
+	netClient, err := getNetworkClient(ctx, c, secretRef)
+	if err != nil {
+		return nil, err
+	}
+	nads := make([]networkv1.NetworkAttachmentDefinition, len(nadRef))
+	for i, ref := range nadRef {
+		// there could be two types of net-attach-def:
+		// pre-created ones for shared resources, e.g. storage network, that each tenant needs to attach
+		// or dynamically created ones on per tenant basis for network isolation.
+		// for pre-created network, simple get and use it. For tenant network, create it using the CNI config, and tenant network annotation.
+		nad, err := netClient.NetworkAttachmentDefinitions(ref.NetAttachDefNamespace).Get(ref.NetAttachDefName, metav1.GetOptions{})
+		if k8serrors.IsAlreadyExists(err) {
+			nads[i] = *nad
+		} else {
+			// if net-attach-def doesn't exist, create one
+			if ref.Config != nil {
+				if nad, err := netClient.NetworkAttachmentDefinitions(ref.NetAttachDefNamespace).Create(
+					&networkv1.NetworkAttachmentDefinition{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: networkName(*ref.Config),
+							Annotations: map[string]string{
+								TenantNetworkAnnotation: "true",
+							},
+						},
+						Spec: networkv1.NetworkAttachmentDefinitionSpec{
+							Config: *ref.Config,
+						},
+					}); err != nil {
+					nads[i] = *nad
+				}
+			}
+		}
+	}
+	return nads, nil
+}
+
+// DeleteTenantNetworkIfExist deletes NetworkAttachmentDefinition annotated as tenant network.
+func DeleteTenantNetworkIfExist(ctx context.Context, c client.Client, secretRef corev1.SecretReference, nads []networkv1.NetworkAttachmentDefinition) error {
+	if len(nads) == 0 {
+		return nil
+	}
+	netClient, err := getNetworkClient(ctx, c, secretRef)
+	if err != nil {
+		return err
+	}
+	for _, nad := range nads {
+		if tenant, ok := nad.ObjectMeta.Annotations[TenantNetworkAnnotation]; ok && tenant == "true" {
+			deleteErr := netClient.NetworkAttachmentDefinitions(nad.ObjectMeta.Namespace).Delete(nad.ObjectMeta.Name, &metav1.DeleteOptions{})
```

Comment:
I would not rely on an annotation to find the NADs that should be deleted. Instead, I would generate the NAD names in the same way as during the reconciliation and delete only these. 
Also, there should be deletion of NADs during reconciliation as well. If there are NADs that are no longer used by VMIs (this could happen if e.g. the config changed and a new NAD was created, and the machines using the old NAD have meanwhile been deleted) they should be deleted at this point.